README
This package has the Vyatta DHCP-server-level configuration templates and scripts.


debian/changelog
vyatta-cfg-dhcp-server (0.12.36+vyos2+lithium7) unstable; urgency=medium

  * Sync with current.

 -- Daniil Baturin <daniil@baturin.org>  Sun, 24 Jun 2018 12:05:34 +0200

vyatta-cfg-dhcp-server (0.12.36+vyos2+current3) unstable; urgency=medium

  * Some fixup on init script.

 -- Mihail Vasilev <mick@corp.linkintel.ru>  Wed, 24 May 2017 16:37:00 +0300

vyatta-cfg-dhcp-server (0.12.36+vyos2+current2) unstable; urgency=medium

  * Use ISC DHCP rather than the vyatta version.

 -- Daniil Baturin <daniil@baturin.org>  Wed, 27 Jan 2016 07:42:18 -0500

vyatta-cfg-dhcp-server (0.12.36+vyos2+current1) unstable; urgency=medium

  * New branch

 -- Kim Hagen <kim.sidney@gmail.com>  Sun, 24 Jan 2016 14:59:48 -0500

vyatta-cfg-dhcp-server (0.12.36+vyos2+lithium6) unstable; urgency=low

  [ Alex Harpin ]
  * vyatta-cfg-dhcp-server: switch hostfile-update to typeless and rename
  * vyatta-cfg-dhcp-server: tidy up "add dynamic hostfile-updates"
  * vyatta-cfg-dhcp-server: don't update /etc/hosts with illegal hostname
  * vyatta-cfg-dhcp-server: formatting changes for style consistency

 -- Alex Harpin <development@landsofshadow.co.uk>  Sun, 22 Nov 2015 19:46:32 +0000

vyatta-cfg-dhcp-server (0.12.36+vyos2+lithium5) unstable; urgency=low

  * vyatta-cfg-dhcp-server: update dh_gencontrol with new development flag

 -- Alex Harpin <development@landsofshadow.co.uk>  Mon, 15 Jun 2015 08:12:59 +0100
 
vyatta-cfg-dhcp-server (0.12.36+vyos2+lithium4) unstable; urgency=low

  [ Brian Hart ]
  * Bug #511: Add dynamic hostfile-updates

 -- Alex Harpin <development@landsofshadow.co.uk>  Sat, 13 Jun 2015 10:06:05 +0100
 
vyatta-cfg-dhcp-server (0.12.36+vyos2+lithium3) unstable; urgency=low

  * Update maintainer address

 -- Alex Harpin <development@landsofshadow.co.uk>  Thu, 25 Dec 2014 13:55:53 +0000
 
vyatta-cfg-dhcp-server (0.12.36+vyos2+lithium2) unstable; urgency=low

  * Force release

 -- Alex Harpin <development@landsofshadow.co.uk>  Mon, 15 Dec 2014 18:20:17 +0000
 
vyatta-cfg-dhcp-server (0.12.36+vyos2+lithium1) unstable; urgency=low

  * New branch

 -- Alex Harpin <development@landsofshadow.co.uk>  Sun, 30 Nov 2014 22:28:25 +0000
 
vyatta-cfg-dhcp-server (0.12.36+vyos1+helium6) unstable; urgency=low

  * vyatta-cfg-dhcp-server: formatting changes for style consistency
  * vyatta-cfg-dhcp-server: formatting changes for style consistency
  * vyatta-cfg-dhcp-server: formatting changes for style consistency
  * vyatta-cfg-dhcp-server: don't prefix hostname with use-host-decl-names
  
 -- Alex Harpin <development@landsofshadow.co.uk>  Fri, 24 Oct 2014 17:53:35 +0100

vyatta-cfg-dhcp-server (0.12.36+vyos1+helium5) unstable; urgency=low

  [ Kim Hagen ]
  * Bug #292 - Inconsistent dhcpdv6.leases location

  [ Alex Harpin ]
  * vyatta-cfg-dhcp-server: prefix host with shared-network-name

  [ Daniil Baturin ]

 -- Daniil Baturin <daniil@baturin.org>  Wed, 08 Oct 2014 16:32:10 +0200

vyatta-cfg-dhcp-server (0.12.36+vyos1+helium4) unstable; urgency=low

  [ Geoff Wright ]
  * Bug #188: Allow domain-search option for dhcp-server

  [ Daniil Baturin ]

 -- Daniil Baturin <daniil@baturin.org>  Tue, 05 Aug 2014 16:27:12 +0000

vyatta-cfg-dhcp-server (0.12.36+vyos1+helium3) unstable; urgency=low

  * Bug #87: Add allowed values completion for "authoritative
    enable/disable" option.

 -- Daniil Baturin <daniil@baturin.org>  Wed, 05 Mar 2014 23:28:19 +0100

vyatta-cfg-dhcp-server (0.12.36+vyos1+helium2) unstable; urgency=low

  * Bug #122: Move DHCPv6 leases file to /config to make it survive
    image upgrades.

 -- Daniil Baturin <daniil@baturin.org>  Wed, 15 Jan 2014 05:16:30 +0100

vyatta-cfg-dhcp-server (0.12.36+vyos1+helium1) unstable; urgency=low

  * New branch

 -- Daniil Baturin <daniil@baturin.org>  Wed, 15 Jan 2014 04:49:00 +0100

vyatta-cfg-dhcp-server (0.12.36+hydrogen1) unstable; urgency=low

  * New branch

 -- Daniil Baturin <daniil@baturin.org>  Sun, 17 Nov 2013 00:05:46 +0100

vyatta-cfg-dhcp-server (0.12.36+daisy2) unstable; urgency=low

  * Bug #4278: allow mapping multiple MAC addresses to the same IP
    address

 -- Daniil Baturin <daniil.baturin@vyatta.com>  Tue, 12 Mar 2013 02:48:47 -0700

vyatta-cfg-dhcp-server (0.12.36+daisy1) unstable; urgency=low

  * create daisy branch

 -- John Southworth <john.southworth@vyatta.com>  Sat, 13 Oct 2012 13:30:26 -0700

vyatta-cfg-dhcp-server (0.12.36) unstable; urgency=low

  * new branch

 -- John Southworth <john.southworth@vyatta.com>  Fri, 12 Oct 2012 19:46:39 -0700

vyatta-cfg-dhcp-server (0.12.35) unstable; urgency=low

  * Update .gitignore
  * Move dhcpd.leases to /config so it is persistent

 -- James Davidson <james.davidson@vyatta.com>  Mon, 17 Sep 2012 19:04:39 -0700

vyatta-cfg-dhcp-server (0.12.34) unstable; urgency=low

  * Bugfix 8293: Apply workaround for dhcp specifically for pacifica.
    General fix to come later

 -- John Southworth <john.southworth@vyatta.com>  Wed, 05 Sep 2012 14:51:38 -0700

vyatta-cfg-dhcp-server (0.12.33) unstable; urgency=low

  * Workaround for static-mapping-parameters filename with double
    quotes, as per comment #14 of bug 5388 - "change the feature's
    configuration

 -- Arthur Xiong <arthur@vyatta.com>  Tue, 24 Apr 2012 17:16:44 -0700

vyatta-cfg-dhcp-server (0.12.32) unstable; urgency=low

  * new branch

 -- Deepti Kulkarni <deepti@vyatta.com>  Sat, 03 Mar 2012 02:25:24 -0800

vyatta-cfg-dhcp-server (0.12.31) unstable; urgency=low

  * Bugfix 6223: Generate correct config file option for static
    mappings.

 -- Bob Gilligan <gilligan@vyatta.com>  Tue, 27 Dec 2011 18:00:10 -0800

vyatta-cfg-dhcp-server (0.12.30) unstable; urgency=low

  * new branch

 -- Deepti Kulkarni <deepti@vyatta.com>  Thu, 07 Jul 2011 20:55:06 -0700

vyatta-cfg-dhcp-server (0.12.29) unstable; urgency=low

  * change for new commit

 -- An-Cheng Huang <ancheng@vyatta.com>  Tue, 10 May 2011 09:20:00 +0800

vyatta-cfg-dhcp-server (0.12.28) unstable; urgency=low

  * Bugfix 6732: Fixed more API issues.


gen-templates.pl
#!/usr/bin/perl

# Module: gen-templates.pl
#
# **** License ****
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License version 2 as
# published by the Free Software Foundation.
#
# This program is distributed in the hope that it will be useful, but
# WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
# General Public License for more details.
#
# A copy of the GNU General Public License is available as
# `/usr/share/common-licenses/GPL' in the Debian GNU/Linux distribution
# or on the World Wide Web at `http://www.gnu.org/copyleft/gpl.html'.
# You can also obtain it by writing to the Free Software Foundation,
# Free Software Foundation, Inc., 51 Franklin St, Fifth Floor, Boston,
# MA 02110-1301, USA.
#
# This code was originally developed by Vyatta, Inc.
# Portions created by Vyatta are Copyright (C) 2010 Vyatta, Inc.
# All Rights Reserved.
#
# Author: Bob Gilligan
# Date: March 2010
# Description: Script to generate Vyatta templates from an input file.
#
# **** End License ****

use strict;
use warnings;

use File::Path;
use Getopt::Long;

# Set to 1 to enable debug output.
#
my $debug_flag = 0;

my $source_file = "src/templates.in";
my $generated_templates_subdir = "generated-templates/";

GetOptions(
    "debug"     => \$debug_flag,
    "source-file=s" =>  \$source_file,
);

sub log_msg {
    my $message = shift;

    print "DEBUG: $message" if $debug_flag;
}


# Main section

if (! open(TPL_SRC, "<$source_file")) {
    printf("Can't open template source file: $source_file\n");
    exit 1;
}

my $tpl_file_open = 0;

my $line;
while ($line = <TPL_SRC>) {
    chomp($line);

    if ($line =~ /^file: /) {
	my $tpl_filename = $line;
	$tpl_filename =~ s/^file: //;
	
	$tpl_filename = $generated_templates_subdir . $tpl_filename;
	my $tpl_dirs = $tpl_filename;
	$tpl_dirs =~ s/\/node.def$//;

	log_msg("Creating directory: $tpl_dirs\n");
	mkpath($tpl_dirs);
	# Check error returns?

	log_msg("Opening template file: $tpl_filename\n");
	if (! open(TPL_FILE, ">$tpl_filename")) {
	    printf("Can't open template file: $tpl_filename\n");
	    exit 1;
	}

	$tpl_file_open++;
    } elsif ($line =~ /^\#\#/) {
	next;
    } else {
	if ($tpl_file_open > 0) {
	    printf(TPL_FILE "$line\n");
	}
    }
}

printf("Wrote $tpl_file_open template files.\n");


scripts/system/on-dhcp-event.sh
#!/bin/bash

# This script came from ubnt.com forum user "bradd" in the following post
# http://community.ubnt.com/t5/EdgeMAX/Automatic-DNS-resolution-of-DHCP-client-names/td-p/651311
# It has been modified by Ubiquiti to update the /etc/host file 
# instead of adding to the CLI.
# Thanks to forum user "itsmarcos" for bug fix & improvements
# Thanks to forum user "ruudboon" for multiple domain fix
# Thanks to forum user "chibby85" for expire patch and static-mapping

if [ $# -lt 5 ]; then
    echo Invalid args
    logger -s -t on-dhcp-event "Invalid args \"$@\""
    exit 1 
fi

action=$1
client_name=$2
client_ip=$3
client_mac=$4
domain=$5
file=/etc/hosts
changes=0

if [ "$domain" == "..YYZ!" ]; then
    client_fqdn_name=$client_name
    client_search_expr=$client_name
else
    client_fqdn_name=$client_name.$domain
    client_search_expr="$client_name\\.$domain"
fi

case "$action" in
    commit) # add mapping for new lease
        echo "- new lease event, setting static mapping for host "\
         "$client_fqdn_name (MAC=$client_mac, IP=$client_ip)"
        if ! [[ "$client_fqdn_name" =~ [[:space:]] ]]; then
            # grep fails miserably with \t in the search expression.
            # In the following line one <Ctrl-V> <TAB> is used after $client_search_expr
            # followed by a single space
            grep -q " $client_search_expr	 #on-dhcp-event " $file
            if [ $? == 0 ]; then
                echo pattern found, removing
                wc1=`cat $file | wc -l`
                sudo sed -i "/ $client_search_expr\t #on-dhcp-event /d" $file
                wc2=`cat $file | wc -l`
                if [ "$wc1" -eq "$wc2" ]; then
                    echo No change
                fi
            else
                echo pattern NOT found
            fi
    
            # check if hostname already exists (e.g. a static host mapping)
            # if so don't overwrite
            grep -q " $client_search_expr	 " $file
            if [ $? == 0 ]; then
                echo host $client_fqdn_name already exists, exiting
                exit 1
            fi

            line="$client_ip\t $client_fqdn_name\t #on-dhcp-event $client_mac"
            sudo sh -c "echo -e '$line' >> $file"
            ((changes++))
            echo Entry was added
        else
            echo "Ignoring illegal hostname (contains spaces)"
        fi
    ;;

    release) # delete mapping for released address
        echo "- lease release event, deleting static mapping for host $client_fqdn_name"
        wc1=`cat $file | wc -l`
        sudo sed -i "/ $client_search_expr\t #on-dhcp-event /d" $file   
        wc2=`cat $file | wc -l`
        if [ "$wc1" -eq "$wc2" ]; then
            echo No change
        else
            echo Entry was removed
            ((changes++))      
        fi
    ;;

    *)
        logger -s -t on-dhcp-event "Invalid command \"$1\""
        exit 1;
    ;;
esac

if [ $changes -gt 0 ]; then
  echo Success
  pid=`pgrep pdns_recursor`
  if [ -n "$pid" ]; then
     sudo rec_control reload-zones
  fi
else
    echo No changes made
fi
exit 0


scripts/system/dhcpdv6-config.pl
#!/usr/bin/perl

# Module: dhcpdv6-config.pl
#
# **** License ****
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License version 2 as
# published by the Free Software Foundation.
#
# This program is distributed in the hope that it will be useful, but
# WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
# General Public License for more details.
#
# A copy of the GNU General Public License is available as
# `/usr/share/common-licenses/GPL' in the Debian GNU/Linux distribution
# or on the World Wide Web at `http://www.gnu.org/copyleft/gpl.html'.
# You can also obtain it by writing to the Free Software Foundation,
# Free Software Foundation, Inc., 51 Franklin St, Fifth Floor, Boston,
# MA 02110-1301, USA.
#
# This code was originally developed by Vyatta, Inc.
# Portions created by Vyatta are Copyright (C) 2010 Vyatta, Inc.
# All Rights Reserved.
#
# Author: Bob Gilligan
# Date: March 2010
# Description: Script to setup DHCPv6 server
#
# **** End License ****

use strict;
use lib "/opt/vyatta/share/perl5/";

use Getopt::Long;
use Vyatta::Config;

# Globals
my $debug_flag;
my $config_filename = "/opt/vyatta/etc/dhcpdv6.conf";
my $op_mode_flag;	# Indicates we are running in op mode if set.

GetOptions(
    "debug"		=> \$debug_flag,
    "config-file=s"	=> \$config_filename,
    "op_mode"		=> \$op_mode_flag,
);

sub log_msg {
    my $message = shift;
    print "DEBUG: $message" if $debug_flag;
}

# Return true if user's param string $param matches
# $match item for item, allowing wild-card string
# in $match.
#
sub param_match {
    my ($match, $param) = @_;
    my @match_array = split(" ", $match);
    my @param_array = split(" ", $param);

    if (scalar(@match_array) != scalar(@param_array)) {
        # match and param arrays don't even have the same number
        # of members.  Can't match.
        return 0;
    }

    my $index = 0;
    foreach my $match_item (@match_array) {
        my $param_item = @param_array[$index];
        $index++;
        if (!($match_item eq "*") && !($match_item eq $param_item)) {
            # no wildcard or exact match
            return 0;
        }
    }
    return 1;
}

# Substitue references of the form $1, $2, etc. in $string with
# values in $param_template string array.  $param_template is the
# Vyatta parameter string that the user configured.
#
sub param_substitute {
    my ($input_string, $param_template) = @_;
    # Turn $param_template into an array of items so that we can
    # reference individual items by number.
    my @param_template_array = split(" ", $param_template);
    my $index = 1;
    
    foreach my $param (@param_template_array) {
        if ($input_string =~ m/VAR-${index}/) {
            log_msg("param_substitue: substituting $param for VAR-${index} \n");
            $input_string =~ s/VAR-${index}/$param/;
        }
        $index++;
    }
    return $input_string;
}

#
# Functions that are used in the "action arrays"
#
my @temp_list = ();

sub write_cf {
    my ($string) = @_;
    printf(CONF_FILE "$string");
}

#
# A simple list is written out to the config file with each item
# separated by commas.
#
sub write_list {
    my ($string) = @_;
    my $num_items = scalar(@temp_list);
    
    if ($num_items > 0) {
        printf(CONF_FILE "$string ");
        my $item_count = 0;
        foreach my $item (@temp_list) {
            if ($item_count > 0) {
                printf(CONF_FILE ", ");
            }
            printf(CONF_FILE "$item");
            $item_count++;
        }
        printf(CONF_FILE ";\n");
    }
    @temp_list = ();
}

# A domain list differs from a simple list in that each
# element must be enclosed in double-quotes, then
# separated by commas.
#
sub write_domain_list {
    my ($string) = @_;
    my $num_items = scalar(@temp_list);

    if ($num_items > 0) {
        printf(CONF_FILE "$string ");
        my $item_count = 0;
        foreach my $item (@temp_list) {
            if ($item_count > 0) {
                printf(CONF_FILE ", ");
            }
            printf(CONF_FILE "\"$item\"");
            $item_count++;
        }
        printf(CONF_FILE ";\n");
    }
    @temp_list = ();
}

sub push_list {
    my ($string) = @_;
    push(@temp_list, $string);
}

#
# We have one "action array" for each of the three transitions:
# Pushing to a new non-leaf level, reach a leaf node with a value, and
# poping back to non-lef level.  Each entry in an action array has
# three elements:  1) a string to match against the parameter string;
# 2) A function to call if it matches, and 3) a string to pass to that
# function.  The first string supports the use of "*" as a wildcard match.
# The third string supports a variable substitution syntax that
# substitues a value from the user's parameter string into the
# string to be passed into the function.
#

my @push_arr = (
    [ "shared-network-name *", \&write_cf, "shared-network VAR-2 {\n" ],
    [ "shared-network-name *", \&write_cf, "    set PoolName = \"VAR-2\";\n" ],
    [ "shared-network-name * subnet *", \&write_cf, "    subnet6 VAR-4 {\n" ],
    [ "shared-network-name * subnet * static-mapping *", \&write_cf, "        host VAR-6 {\n" ],
    [ "shared-network-name * subnet * address-range prefix *", \&write_cf, "        range6 VAR-7" ],
    [ "shared-network-name * subnet * address-range prefix * temporary", \&write_cf, " temporary" ],
);

my @pop_arr = (
    [ "shared-network-name *", \&write_cf, "}\n" ],
    [ "shared-network-name * subnet *", \&write_cf, "    }\n" ],
    [ "shared-network-name * subnet * name-server", \&write_list, "        option dhcp6.name-servers" ],
    [ "shared-network-name * subnet * domain-search", \&write_domain_list, "        option dhcp6.domain-search" ],
    [ "shared-network-name * subnet * sip-server-address", \&write_list, "        option dhcp6.sip-servers-addresses" ],
    [ "shared-network-name * subnet * sip-server-name", \&write_domain_list, "        option dhcp6.sip-servers-names" ],
    [ "shared-network-name * subnet * nis-server", \&write_list, "        option dhcp6.nis-servers" ],
    [ "shared-network-name * subnet * nisplus-server", \&write_list, "        option dhcp6.nisp-servers" ],
    [ "shared-network-name * subnet * sntp-server", \&write_list, "        option dhcp6.sntp-servers" ],
    [ "shared-network-name * subnet * address-range start *", \&write_cf, ";\n" ],
    [ "shared-network-name * subnet * address-range prefix *", \&write_cf, ";\n" ],
    [ "shared-network-name * subnet * static-mapping *", \&write_cf, "        }\n" ],
);

my @leaf_arr = (
    [ "preference *", \&write_cf, "option dhcp6.preference VAR-2;\n" ],
    [ "shared-network-name * subnet * name-server *", \&push_list, "VAR-6" ],
    [ "shared-network-name * subnet * domain-search *", \&push_list, "VAR-6" ],
    [ "shared-network-name * subnet * sip-server-address *", \&push_list, "VAR-6" ],
    [ "shared-network-name * subnet * sip-server-name *", \&push_list, "VAR-6" ],
    [ "shared-network-name * subnet * nis-server *", \&push_list, "VAR-6" ],
    [ "shared-network-name * subnet * nisplus-server *", \&push_list, "VAR-6" ],
    [ "shared-network-name * subnet * nis-domain *", \&write_cf, "        option dhcp6.nis-domain-name \"VAR-6\";\n" ],
    [ "shared-network-name * subnet * nisplus-domain *", \&write_cf, "        option dhcp6.nisp-domain-name \"VAR-6\";\n" ],
    [ "shared-network-name * subnet * sntp-server *", \&push_list, "VAR-6" ],
    [ "shared-network-name * subnet * lease-time maximum *", \&write_cf, "        max-lease-time VAR-7;\n" ],
    [ "shared-network-name * subnet * lease-time minimum *", \&write_cf, "        min-lease-time VAR-7;\n" ],
    [ "shared-network-name * subnet * lease-time default *", \&write_cf, "        default-lease-time VAR-7;\n" ],
    [ "shared-network-name * subnet * address-range start * stop *", \&write_cf, "        range6 VAR-7 VAR-9" ],
    [ "shared-network-name * subnet * static-mapping * ipv6-address *", \&write_cf, "            fixed-address6 VAR-8;\n" ],
    [ "shared-network-name * subnet * static-mapping * identifier *", \&write_cf, "            host-identifier option dhcp6.client-id VAR-8;\n" ],
);

#
# Walk through the action array passed in by reference. If an entry is
# found in that array whose first string passes the users's parameter
# string passed in, then perform parameter substitution on the third
# string in the entry, and call the function identified by the second
# element of the entry.
#
sub action_func {
    my ($action_arr_ref, $param) = @_;
    my @action_arr = @$action_arr_ref;

    foreach my $row (0 .. scalar(@action_arr) - 1) {
        my $match = $action_arr[$row][0];
        if (param_match ($match, $param)) {
            my $func = $action_arr[$row][1];
            my $arg = $action_arr[$row][2];
            my $action_string = param_substitute($arg, $param);
            &$func($action_string);
        }
    }
}

#
# Recursive walk of the config tree starting at $level. $vc is the
# config pointer.  $depth records the current tree depth, primarily
# for debugging.  The final three args are references to the three
# "action arrays" discussed above.
#
sub walk_tree {
    my ($vc, $level, $depth, $push_arr_ref, $pop_arr_ref, $leaf_arr_ref) = @_;
    my @values;
    
    log_msg("in walk_tree at depth $depth level is: $level \n");
    if ($op_mode_flag) {
        @values = $vc->returnOrigValues($level);
    } else {
        @values = $vc->returnValues($level);
    }

    my $num_values = scalar(@values);
    if ($num_values == 0) {

        # We don't know if leaf nodes are single or multi-valued.
        # The returnOrigValues and returnValues functions only
        # return values if a leaf node is multi-valued.  So we need
        # to check here to see if this is a leaf single-valued node.
        #
        my $val;
        if ($op_mode_flag) {
            $val = $vc->returnOrigValue($level);
        } else {
            $val = $vc->returnValue($level);
        }

        if (defined $val) {
            @values = ( $val );
            $num_values = 1;
        }
    }

    # Note that valueless leaf nodes have no "values", so they don't
    # show up as "leafs" in this test.  So, actions on such nodes must
    # be taken at push time.
    if ($num_values > 0) {
        log_msg("Push to leaf: $level\n");
        action_func($push_arr_ref, $level);

        foreach my $value (@values) {
            my $leaf_value = $level . " " . $value;
            log_msg("Leaf: $leaf_value\n");

            action_func($leaf_arr_ref, $leaf_value);
        }
        log_msg("Pop to leaf: $level\n");
        action_func($pop_arr_ref, $level);
    } else {
        log_msg("Push to: $level\n");
        action_func($push_arr_ref, $level);

        my @node_array;
        if ($op_mode_flag) {
            @node_array = $vc->listOrigNodes($level);
        } else {
            @node_array = $vc->listNodes($level);
        }

        foreach my $node (sort (@node_array)) {
            log_msg("node at depth $depth is $node\n");
            walk_tree ($vc, $level . " " . $node, $depth + 1, $push_arr_ref, $pop_arr_ref, $leaf_arr_ref);
        }
        log_msg("Pop to: $level\n");
        action_func($pop_arr_ref, $level);
    }
}

#
# Main section
#

my $vcDHCP = new Vyatta::Config();

# Do we need to perform any cross-parameter validation checks?
# This might include:
#  - Subnet prefix overlap
#  - Range overlaps

# Walk the config tree
#
my $exists;
if ($op_mode_flag) {
    $exists = $vcDHCP->existsOrig('service dhcpv6-server');
} else {
    $exists = $vcDHCP->exists('service dhcpv6-server');
}

if ($exists) {
    # Open the config file
    #
    if (! open(CONF_FILE, ">$config_filename")) {
        printf("Can't open config file for writing: $config_filename\n");
        exit 1;
    }

    printf("Generating the DHCPv6 config file...\n");
    #
    # Write some comments so people know where it came from.
    #
    printf(CONF_FILE "# This file is auto-generated by the Vyatta configuration sub-system.\n");
    printf(CONF_FILE "# Do not edit it by hand.\n");
    my $iam = `whoami`;
    chomp ($iam);
    printf(CONF_FILE "# Auto-generated by: $iam\n");
    my $date_time = `date`;
    chomp ($date_time);
    printf(CONF_FILE "# Auto-generated on: $date_time\n");
    printf(CONF_FILE "#\n");

    $vcDHCP->setLevel('service dhcpv6-server');
    log_msg("Initial call to walk_tree.\n");
    walk_tree($vcDHCP, "", 0, \@push_arr, \@pop_arr, \@leaf_arr);
    #
    # Close the config file, we're done!
    #
    close(CONF_FILE);
} else {
    if ($op_mode_flag) {
        printf("DHCPv6 server is not configured.\n");
    }
    # No error message required when run in config mode since
    # it is expected behavior.
    exit 1;
}

exit 0;
